internal/state/backend_refs.go

```diff
@@ -0,0 +1,156 @@
+package state
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/resolver"
+)
+
+type (
+	// ruleIndex is the index of the HTTPRouteRule.
+	ruleIndex int
+	// resolvedBackends is a map of backend names to their endpoints.
+	resolvedBackends map[string][]resolver.Endpoint
+	// backendGroupsByRule is a map of rule indexes to backend groups.
+	backendGroupsByRule map[ruleIndex]BackendGroup
```

Comment:
Can our indexes be non-contiguous? It doesn't look like it from how it's constructed. If the key is a contiguous integer this can be a slice and referenced in the same manner.

Reply:
This can be a slice. I chose a map because I didn't want to worry about index out of range panics. But those shouldn't occur unless we mess something up.

Reply:
Good question. I'd say a subtle bug is worse. That convinces me that this should be a slice.

Reply:
At the worst, I can see a small API to safely access slices. It's also more efficient accessing, iterating, and storage.